src/zon.zig

```diff
@@ -0,0 +1,661 @@
+const std = @import("std");
+const Zcu = @import("Zcu.zig");
+const Sema = @import("Sema.zig");
+const InternPool = @import("InternPool.zig");
+const Type = @import("Type.zig");
+const Zir = std.zig.Zir;
+const AstGen = std.zig.AstGen;
+const CompileError = Zcu.CompileError;
+const Ast = std.zig.Ast;
+const Allocator = std.mem.Allocator;
+const assert = std.debug.assert;
+const File = Zcu.File;
+const LazySrcLoc = Zcu.LazySrcLoc;
+const Ref = std.zig.Zir.Inst.Ref;
+const NullTerminatedString = InternPool.NullTerminatedString;
+const NumberLiteralError = std.zig.number_literal.Error;
+
+const LowerZon = @This();
+
+sema: *Sema,
+file: *File,
+file_index: Zcu.File.Index,
+
+/// Lowers the given file as ZON. `res_ty` is a hint that's used to add indirection as needed to
+/// match the result type, actual type checking is not done until assignment.
```

Comment:
> `actual type checking is not done until assignment`

This is kind of problematic for two reasons:
* It leads to less useful error messages. Coercing as we lower will give more helpful type errors since the error location will be the specific sub-expression with an incorrect type.
* More importantly: once we remove anonymous struct types (#16865), coercing structs through pointers -- which this presumably relies on -- won't work.